In [ ]:
# default_exp linear_regression

In [ ]:
#hide
from nbdev.showdoc import *

# Linear Regression
> This tutorial describes how to fit a curve using linear regression.

## Regression


We wish to learn $f:X\to Y$, where $Y$ is a real number, given $\{<x^1,y^1>,\cdots, <x^m,y^m>\}$.

**Approach**

1- Choose some parametraized form for $P(Y|X;\theta)$

2- Derive learning algorithms as Maximum Likelihood Estimates (MLE), or Maximum a Posteriori (MAP) estimate for $\theta$

## Simple Linear Regression Model

Let's assume that we are going to predict the price of a house, $y$, based on its square feet, $x$. Therefore, training set = $\{<x^1,y^1>,\cdots,<x^m,y^m>\}$, where:

- $m$: number of trainig examples

- $x$: input variable

- $y$: output variable or target label

- $(x^i,y^i)$: is the $i$th training example

<img src="images/linear_regression_ex.png" width="500" height="500" />

We would like to find a **line** that **fits** our training examples the **best**.

<img src="images/linear_regression_ex_model.png" width="500" height="500"/>

Since we are looling for a line to fit our training set, we need to find a function $\hat y = f(x)$ that estimates $y^i$ for all $1\leq i \leq m$. Thus, we have:

$\hat y = f(x) = w_0 + w_1x$

$w_0$ and $w_1$ are the paramaters that we need to find.

## Cost Function/Error Function

We wish to estimate the actual $y$, i.e. we want to minimize the error as much as possible. The error is the difference between the actual $y^i$ and our estimated/predicted $\hat{y}^i$ for all $1\leq i \leq m$. 

The error/cost function is:






$$J(w_0,w_1) = \frac{1}{m}\sum_{i=1}^{m}\left( \hat{y}^i - (w_0 + w_1 x^i)\right)^2$$

We get the average of the function squared to make the arithmatic easier.

Precisely, we need to minimize the **mean residual squared error**:

$$\text{minimize } J(w_0,w_1) = \underset{\mathbf{w0,w_1}}{\arg\min} \frac{1}{m}\sum_{i=1}^{m}\left( \hat{y}^i - (w_0 + w_1 x^i)\right)^2$$


## Fitting the Linear Regression Model using Gradient Descent Algorithm


In order to mininze the cost function, we need to take the gradient (i.e. derivative) of the function with resptect to our parameters $w_0$ and $w_1$, set it zero and solve for $w_0$ and $w_1$.
 
 
$$\frac{\partial{J}}{\partial{w_0}} = \frac{-2}{m} \sum_{i=1}^{m}\left(\hat{y}^i - (w_0 + w_1 x^i)\right)$$

$$\frac{\partial{J}}{\partial{w_1}} = \frac{-2}{m} \sum_{i=1}^{m}x^i\left(\hat{y}^i - (w_0 + w_1 x^i)\right)$$

### The gradient descent algorithm:

initialize $w_0^{(0)} = w_1^{(0)} = 0$, for $t=0$

for $t=1$ to *num_iterations*
   

   
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
$w_0^{(t+1)} = w_0^{(t)} - \eta \frac{-2}{m} \sum_{i=1}^{m}\left(\hat{y}^i - (w_0 + w_1 x^i)\right) $

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
$w_1^{(t+1)} = w_1^{(t)} - \eta\frac{-2}{m} \sum_{i=1}^{m}x^i\left(\hat{y}^i - (w_0 + w_1 x^i)\right)$

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$t = t+1$
    

In [ ]:
# export
import numpy as np
import pandas as pd
import altair as alt
def generate_data():
    """It generates dummy data."""
    noise = np.random.randn(100,1)
    X = 2 * np.random.rand(100,1)
    y = 5 + 3 * X + noise
    return X,y

In [ ]:
# export
def make_point_plot(x,y):
    """It plots the point chart of the data"""
    
    data_points = pd.DataFrame({'x': x.flatten(), 'y': y.flatten()})
    chart = alt.Chart(data_points).mark_point(size=50, color='red',filled=True).encode(
        x="x",
        y="y"
    )
    return chart
    

In [ ]:
# export
def make_line_plot(x,y):
    """It plots the line chart of the data"""
    
    data = pd.DataFrame({'x': x.flatten(), 'y': y.flatten()})
    line = alt.Chart(data).mark_line(size=3).encode(
        x="x",
        y="y"
    )
    return line

In [ ]:
#export
def gradient_descent(data,w_0_t,w_1_t,learning_rate,num_iterations):
    """Gradient descent implementation, which gets `data`, starting `w_0` and `w_1`, `learning_rate` 
    and the number of iterations `num_iterations`"""
    
    w_0 = 0
    w_1 = 0
    (X,y) = data
    N = len(X)
    w_0_t = 0
    w_1_t = 0
    for t in range(0,num_iterations):
        w_0_deriv = np.zeros((N,N))
        w_1_deriv = np.zeros((N,N))
        w_0_deriv = -2 * (y - (w_0_t + w_1_t * X))
        w_1_deriv = -2 * np.dot(X.T, (y - (w_0_t + w_1_t * X)))
        w0_sum = np.sum(w_0_deriv,axis=0)
        w1_sum = np.sum(w_1_deriv,axis=0)
        w_0 = w_0 - learning_rate * (w0_sum / N)
        w_1 = w_1 - learning_rate * (w1_sum / N)
        w_0_t = w_0
        w_1_t = w_1
    return w_0, w_1

## Run an Example

In this example, we generate some linear-looking data and then find the line that fits the data. The function that we use to generate the data is $y=5+3x+\text{Gaussian noise}$. Our goal is to find $\theta=[w_0,w_1]$ where $w_0=5$ and $w_1=3$ or close enough, as we have noise and it makes it impossible to recover the exact paratmeters of the function. We use the `generate_data` function to generate the training data. 

In [ ]:
X,y = generate_data()

The chart below shows the generated dataset.

In [ ]:
origin_chart = make_point_plot(X,y)

#show the chart
origin_chart

alt.Chart(...)

We set the initial parameters to 0:  $w_0^{(0)} = w_1^{(0)} = 0$, define the number of iterations of the graditent descent algorithm as well as the learning rate. Then, we run the `gradient_descent` function. The outputs of this function are the estimated parameters $w_0$ and $w_1$.

In [ ]:
initial_w_0 = 0
initial_w_1 = 0
num_iterations = 5000
learning_rate = 0.001

# X,y = generate_data()
data = (X,y)
w_0,w_1 = gradient_descent(data,initial_w_0,initial_w_1,learning_rate,num_iterations)

print("w_0 = {}, w_1 = {}".format(w_0,w_1))

w_0 = [4.65664658], w_1 = [3.33942319]


The estimated parameters are $w_0 = 4.656$ and $w_1 = 3.339$ that are close enough to $w_0=5$ and $w_1=3$. Let's plot the line.

In [ ]:
y_predicted = w_0 + w_1 * X
line = make_line_plot(X,y_predicted)

#show the charts
origin_chart + line

alt.LayerChart(...)

In [ ]:
# hide
initial_w_0 = 0
initial_w_1 = 0
num_iterations = 5000
l_rates = [0.001, 0.1,0.2]
learning_rate = 0.001
all_params = []

X,y = generate_data()
data = (X,y)
for i in range(0,len(l_rates)):
    learning_rate = l_rates[i]
    w_0,w_1 = gradient_descent(data,initial_w_0,initial_w_1,learning_rate,num_iterations)
    all_params.append((w_0,w_1))
    print("learning_rate: {:.4f} => w_0 = {}, w_1 = {}".format(learning_rate,w_0,w_1))

learning_rate: 0.0010 => w_0 = [4.65664658], w_1 = [3.33942319]
learning_rate: 0.1000 => w_0 = [4.90858171], w_1 = [3.11876371]
learning_rate: 0.2000 => w_0 = [4.90858171], w_1 = [3.11876371]
